src/main/java/frc/robot/commands/testing/TestSubsystem.java

```diff
@@ -0,0 +1,171 @@
+        }
+    }
+
+    @Override
+    public void initialize() {
+        for (int i = 0; i < motors.length; i++) {
+            startingPositions[i] = encoders[i].getPosition();
+            motors[i].set(toTest.getMotorTestSpeed());
+        }
+
+        for (HashMap.Entry<String, Boolean> set : toTest.initCustomTests().entrySet()) {
+            testStates.put(set.getKey(), false);
+        }
+
+        timer.start();
+    }
+
+    @Override
+    public void execute() {
+        for (int i = 0; i < motors.length; i++) {
+            if (encoders[i].getPosition() > startingPositions[i] + toTest.getMotorTestRotations()) {
+                motors[i].set(0);
+                testStates.replace("Motor " + motors[i].getDeviceId(), true);
+            }
+        }
+
+        for (int i = 0; i < colorSensors.length; i++) {
+            if (colorSensors[i].getLoopTimeMilliseconds() < toTest.getMaxSensorLoopMilliseconds()
+                && colorSensors[i].getProximity() > 0
+                && colorSensors[i].getColorString() != "Unknown") {
+                testStates.replace("Color Sensor " + i, true);
+            } else {
+                testStates.replace("Color Sensor " + i, false);
+            }
+        }
+
+        for (int i = 0; i < lidars.length; i++) {
+            if (lidars[i].getLoopTimeMilliseconds() < toTest.getMaxSensorLoopMilliseconds()
+                && lidars[i].getDistance() > 0) {
+                testStates.replace("Lidar " + i, true);
+            } else {
+                testStates.replace("Lidar " + i, false);
+            }
+        }
+
+        for (int i = 0; i < limelights.length; i++) {
+            // TODO - What else can we test on limelights?
+            if (limelights[i] != null) {
+                testStates.replace("Limelight " + i, true);
+            } else {
+                testStates.replace("Limelight " + i, false);
+            }
+        }
+
+        testStates.putAll(toTest.runCustomTests());
+    }
+
+    @Override
+    public void end(final boolean interrupted) {
+        timer.stop();
+
+        for (int i = 0; i < motors.length; i++) {
+            motors[i].set(0);
+        }
+    }
+
+    @Override
+    public boolean isFinished() {
+        boolean isSuccessful = true;
+
+        for (HashMap.Entry<String, Boolean> set : testStates.entrySet()) {
+            isSuccessful &= set.getValue();
+        }
+
+        if (isSuccessful) {
+            ShuffleboardDriver.statusDisplay.addStatus(
+                "Test Status",
+                "Test was successful",
+                Colors.GREEN,
+                5
+            );
+            return true;
+
+        } else if (timer.hasElapsed(toTest.getAllowedTimeSeconds())) {
+            String toPrint = "";
+
+            for (HashMap.Entry<String, Boolean> set : testStates.entrySet()) {
+                if (!set.getValue()) {
+                    toPrint += (" " + toTest.getTestName()
+                        + ": " + set.getKey() + ",");
```

Comment:
You should just have the test name once in the to print. The printed statement is already specific to the subsystem. If two things failed in one subsystem you don't want to print the subsystem name twice.